invisible_cities/reco/deconv_functions.py

```diff
@@ -0,0 +1,288 @@
+import numpy  as np
+import pandas as pd
+
+from typing  import List
+from typing  import Tuple
+from typing  import Optional
+from typing  import Callable
+
+from scipy        import interpolate
+from scipy.signal import fftconvolve, convolve
+
+from ..core .core_functions import shift_to_bin_centers
+from ..core .core_functions import in_range
+
+def cut_and_redistribute_df(cut_condition : str,
+                            variables     : List[str]=[]) -> Callable:
+    '''
+    Apply a cut condition to a dataframe and redistribute the cut out values
+    of a given variable.
+
+    Parameters
+    ----------
+    df      : dataframe to be cut
+
+    Initialization parameters:
+        cut_condition : String with the cut condition (example "Q > 10")
+        variables     : List with variables to be redistributed.
+
+    Returns
+    ----------
+    pass_df : dataframe after applying the cut and redistribution.
+    '''
+    def cut_and_redistribute(df : pd.DataFrame) -> pd.DataFrame:
+        pass_df = df.query(cut_condition)
+        pass_df._is_copy = False
+        with np.errstate(divide='ignore'):
+            for redist_variable in variables:
+                pass_df[redist_variable] = pass_df[redist_variable] * np.divide(df[redist_variable].sum(), pass_df[redist_variable].sum())
+        return pass_df
+
+    return cut_and_redistribute
+
+def drop_isolated_sensors(distance  : List[float]=[10., 10.],
+                          variables : List[str  ]=[        ]) -> Callable:
+    """
+    Drops rogue/isolated hits (SiPMs) from a groupedby dataframe.
+
+    Parameters
+    ----------
+    df      : GroupBy ('event' and 'npeak') dataframe
+
+    Initialization parameters:
+        distance  : Distance to check for other sensros. Usually equal to sensor pitch.
+        variables : List with variables to be redistributed.
+
+    Returns
+    ----------
+    pass_df : hits after removing isolated hits
+    """
+    dist = np.sqrt(distance[0] ** 2 + distance[1] ** 2)
+
+    def drop_isolated_sensors(df : pd.DataFrame) -> pd.DataFrame:
+        x       = df.X.values
+        y       = df.Y.values
+        mask_xy = [False if np.ma.masked_equal(np.sqrt((x - xi)**2 + (y - yi)**2), 0.0, copy=False).min() > dist
+                         else True
+                         for xi, yi in zip(x, y)]
```

Comment:
again here, this function could be of general use to filter 'hits' that have not close neighbors. Maybe a more general name *select_hits_with_neighbors_in_distance*. 

I do not see why we pass (x, y)-distance when, if I understand right, there is one the radius used.

Reply:
Could be only a single distance but I feel it's more intuitive to give the pitch en each dimension. I somehow suspect that if we just write the distance the user may just assume that to keep the first ring is enough with 1 pitch distance instead of the 1.41 * pitch but I do not have a strong opinion here, it can be easily changed.